src/layouting/layouting.jl

```diff
@@ -21,16 +21,96 @@ function layout_text(
     pos = to_ndim(Point3f0, mpos, 0)
     scales = Vec2f0[glyph_scale!(atlas, c, ft_font, rscale) for c in string]
 
-    positions2d = calc_position(string, Point2f0(0), rscale, ft_font, atlas)
-    aoffset = align_offset(Point2f0(0), positions2d[end], atlas, rscale, ft_font, offset_vec)
-    aoffsetn = to_ndim(Point3f0, aoffset, 0f0)
-    positions = map(positions2d) do p
-        pn = rot * (to_ndim(Point3f0, p, 0f0) .+ aoffsetn)
-        pn .+ (pos)
+    glyphpos = glyph_positions(string, ft_font, rscale, offset_vec[1], offset_vec[2];
+        lineheight_factor = 1.5, justification = 0.0)
+
+    positions = Point3f0[]
+    for (i, group) in enumerate(glyphpos)
+        for gp in group
+            p = to_ndim(Point3f0, gp, 0) #./ Point3f0(4, 4, 1)
+            # rotate around the alignment point (this is now at [0, 0, 0])
+            p_rotated = rot * p
+            push!(positions, pos .+ p_rotated) # TODO why division by 4 necessary?
+        end
+        # between groups, push a random point for newline, it doesn't matter
+        # what it is
+        if i < length(glyphpos)
+            push!(positions, Point3f0(0, 0, 0))
+        end
     end
+
     return positions, scales
 end
 
+
+function glyph_positions(str::AbstractString, font, fontscale, halign, valign; lineheight_factor = 1.0, justification = 0.0)
+
+    # this is a countermeasure against Cairo messing with FreeType font pixel sizes
+    # when drawing. We reset them every time which is hacky but seems to work
+    FreeTypeAbstraction.FreeType.FT_Set_Pixel_Sizes(font, 64, 64)
+    FreeTypeAbstraction.FreeType.FT_Set_Transform(font, C_NULL, C_NULL)
+    
+
+    # make lineheight a multiple of font's M height
+    lineheight = inkheight(FreeTypeAbstraction.internal_get_extent(font, 'M')) * lineheight_factor
+
+    lines = split(str, "\n")
+    extents = [[FreeTypeAbstraction.internal_get_extent(font, c) for c in l] for l in lines]
+
+    xkernings = [[FreeTypeAbstraction.kerning(c1, c2, font)[1]
+        for (c1, c2) in zip(chop(l, head = 0, tail = 1), chop(l, head = 1, tail = 0))]
+            for l in lines]
+
+    # add or subtract kernings?
+    xs = [cumsum([0; hadvance.(extgroup[1:end-1]) .+ kerngroup]) for (extgroup, kerngroup) in zip(extents, xkernings)]
+    [hadvance.(extgroup[1:end-1]) for extgroup in extents]
```

Comment:
This line seems unused?

Reply:
true, think I deleted that in the newer versions